src/compiler/program.ts

```diff
@@ -91,6 +148,150 @@ namespace ts {
+                }
+            }
+        }
+        // then read 'types'
+        if (!typesFile && jsonContent.types) {
+            if (typeof jsonContent.types === "string") {
+                fieldName = "types";
+                typesFile = jsonContent.types;
+            }
+            else {
+                if (state.traceEnabled) {
+                    trace(state.host, Diagnostics.Expected_type_of_0_field_in_package_json_to_be_string_got_1, "types", typeof jsonContent.types);
+                }
+            }
+        }
+        if (typesFile) {
+            const typesFilePath = normalizePath(combinePaths(baseDirectory, typesFile));
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.package_json_has_0_field_1_that_references_2, fieldName, typesFile, typesFilePath);
+            }
+            return typesFilePath;
+        }
+        return undefined;
+    }
+
+    function tryLoadTypeDeclarationFile(searchPath: string, failedLookupLocations: string[], state: ModuleResolutionState) {
+        let typesFile: string;
+        const packageJsonPath = combinePaths(searchPath, "package.json");
+        if (state.host.fileExists(packageJsonPath)) {
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.Found_package_json_at_0, packageJsonPath);
+            }
+            typesFile = tryReadTypesSection(packageJsonPath, searchPath, state);
+            if (!typesFile) {
+                if (state.traceEnabled) {
+                    trace(state.host, Diagnostics.package_json_does_not_have_types_field);
+                }
+            }
+        }
+        else {
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.File_0_does_not_exist, packageJsonPath);
+            }
+            failedLookupLocations.push(packageJsonPath);
+        }
+
+        if (!typesFile) {
+            typesFile = "index.d.ts";
+        }
+
+        const combinedPath = normalizePath(combinePaths(searchPath, typesFile));
+        if (state.host.fileExists(combinedPath)) {
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.File_0_exist_use_it_as_a_name_resolution_result, combinedPath);
+            }
+            return combinedPath;
+        }
+        else {
+            if (state.traceEnabled) {
+                trace(state.host, Diagnostics.File_0_does_not_exist, combinedPath);
+            }
+            failedLookupLocations.push(combinedPath);
+            return undefined;
+        }
+    }
+
+    function getEffectiveTypesPrimarySearchPaths(options: CompilerOptions): string[] {
+        return options.typesSearchPaths || defaultLibrarySearchPaths;
+    }
+
+    export function resolveTypeReferenceDirective(typeReferenceDirectiveName: string, containingFile: string, compilationRoot: string, options: CompilerOptions, host: ModuleResolutionHost): ResolvedTypeReferenceDirectiveWithFailedLookupLocations {

```

Comment:
so why not use `loadNodeModuleFromDirectory` here and pass extensions to be `[".d.ts"]` ?